granulate_utils/config_feeder/client/client.py

```diff
@@ -0,0 +1,182 @@
+import asyncio
+import json
+import logging
+from collections import defaultdict
+from typing import Any, DefaultDict, Dict, Optional, Union, cast
+
+from pydantic import BaseModel
+from requests import Session
+from requests.exceptions import ConnectionError, JSONDecodeError
+
+from granulate_utils.config_feeder.client.bigdata import get_node_info
+from granulate_utils.config_feeder.client.exceptions import APIError, ClientError
+from granulate_utils.config_feeder.client.models import CollectionResult, ConfigType
+from granulate_utils.config_feeder.client.yarn.collector import YarnConfigCollector
+from granulate_utils.config_feeder.client.yarn.models import YarnConfig
+from granulate_utils.config_feeder.core.errors import raise_for_code
+from granulate_utils.config_feeder.core.models.aggregation import CreateNodeConfigsRequest, CreateNodeConfigsResponse
+from granulate_utils.config_feeder.core.models.cluster import ClusterCreate, CreateClusterRequest, CreateClusterResponse
+from granulate_utils.config_feeder.core.models.collection import CollectorType
+from granulate_utils.config_feeder.core.models.node import CreateNodeRequest, CreateNodeResponse, NodeCreate, NodeInfo
+from granulate_utils.config_feeder.core.models.yarn import NodeYarnConfigCreate
+
+DEFAULT_API_SERVER_ADDRESS = "https://api.granulate.io/config-feeder/api/v1"
+DEFAULT_REQUEST_TIMEOUT = 3
+
+
+class ConfigFeederClient:
+    def __init__(
+        self,
+        token: str,
+        service: str,
+        *,
+        logger: Union[logging.Logger, logging.LoggerAdapter],
+        server_address: Optional[str] = None,
+        yarn: bool = True,
+        collector_type=CollectorType.SAGENT,
+    ) -> None:
+        if not token or not service:
+            raise ClientError("Token and service must be provided")
+        self.logger = logger
+        self._token = token
+        self._service = service
+        self._cluster_id: Optional[str] = None
+        self._collector_type = collector_type
+        self._server_address: str = server_address.rstrip("/") if server_address else DEFAULT_API_SERVER_ADDRESS
+        self._is_yarn_enabled = yarn
+        self._yarn_collector = YarnConfigCollector(logger=logger)
+        self._last_hash: DefaultDict[ConfigType, Dict[str, str]] = defaultdict(dict)
+        self._init_api_session()
+
+    def collect(self) -> None:
+        if (node_info := get_node_info(self.logger)) is None:
+            self.logger.warning("not a Big Data host, skipping")
+            return None
+
+        collection_result = asyncio.run(self._collect(node_info))
+
+        if self._cluster_id is None and (node_info.is_master or not collection_result.is_empty):
+            self._register_cluster(node_info)
```

Comment:
Why not register always and let the backend decide?

Reply:
up to you, IMO this creates unnecessary load on server and DB.

Reply:
ok